### ISSUE TITLE
Strange behaviour of BufReader

### ISSUE THREAD

**Issue by antekone** [rust-lang/rust#22570]
Edit: It seems I was too hasty about this, and this actually is an expected behaviour. I'm not sure I can remove this entry, so please close it.

On Reddit I've got an advice that this could be a known issue and it will be fixed in time, but I couldn't find it in the issue list, so here it goes.

Here's the code of the POC:

    #![feature(fs)]
    #![feature(io)]
    
    use std::io::prelude::*;
    use std::io::BufReader;
    use std::fs::File;
    
    fn main() {
        check_sample_file();
    }
    
    fn check_sample_file() {
        let mut reader = BufReader::new(File::open("test.bin").unwrap());
    
        let mut buf = [0u8; 520];
        let mut sum_read = 0us;
    
        for i in (0..140us) {
            match reader.read(&mut buf) {
                Ok(count) => {
                    sum_read += count;
                    println!("Read {} bytes, sum of read bytes: {}", count, sum_read);
                },
                Err(_) => return
            }
        }
    }

All it does is that it reads the file in 520-byte chunks by using a buffering layer in the form of BufReader. Now here's the output:

    [...]
    Read 520 bytes, sum of read bytes: 64480
    Read 520 bytes, sum of read bytes: 65000
    Read 520 bytes, sum of read bytes: 65520
    Read 16 bytes, sum of read bytes: 65536
    Read 520 bytes, sum of read bytes: 66056
    Read 520 bytes, sum of read bytes: 66576
    Read 520 bytes, sum of read bytes: 67096
    [...]

Why it suddenly read 16 bytes, up to the boundary of its buffer capacity? Shouldn't BufReader handle this case and abstract me from these kinds of problems?

My Rustc version is: rustc 1.0.0-nightly (dfc5c0f1e 2015-02-18) (built 2015-02-19)

**Comment by antekone** [comment]
Also, I've just checked, and it seems that `std::old_io::BufferedReader` suffers from the same issue.

**Comment by tsezgin** [comment]
Hi Grzegorz, 
I hit the same problem recently, and I wonder what made you convinced that this is not a valid issue.
Althought It seems to be the expected behaviour, I'm not convinced that it is the correct behaviour.
I have not encountered such behaviour in other languages such as (C++, Java, C#). 
And, how did you handle this issue in your application? a wrapper method?
thanks.

**Comment by antekone** [comment]
IIRC, it was mostly due to a comment I've received on a different site. I've made [this thread](https://www.reddit.com/r/rust/comments/2wjaqa/strange_behaviour_of_bufreader_and_offset_65536/) on Reddit. The comment from burntsushi made an interesting reply implying `read()` is too low-level to do any buffering by its own. Also the documentation of `read()` says, that it just "pulls some bytes from the source, returning number of bytes actually read". This convinced me that this indeed is designed behavior (I didn't like it as well).

Currently there are functions like `read_exact()` and `read_to_end()`, have you tried these?

**Comment by tsezgin** [comment]
I had seen read_exact() and read_to_end() functions in the documentation, but had preferred using read(). Now, reading your suggestion, I switched to read_exact() and my application works properly.

But I'm still not convinced about BufReader::read() behaviour, when backend is a File, because it changes behaviour observed by the application.

In my case, initially I started with using File::read(), and the application worked as expected. 
my buffer was always filled, except the last read(), which is EOF case.

Then, just to improve the IO performance, I wrapped the File object with BufReader, and the behaviour of the application has changed. I received partial fills although there are more data in the file.
I would expect such behaviour while reading from from a TCP socket.
But while reading from a file, I would expect BufReader to return full buffers. (by properly refilling its internal buffer).
I believe, Java and C# buffered readers behave this way, without needing a separate read_exact() method.
Anyway, thanks again for your explanations and quick response.
